### ISSUE TITLE
FTP Adapter WINDOWS

### ISSUE THREAD

**Issue by kinncj** [thephpleague/flysystem#451]
It doesn't seems to work properly with `windows` systems, since the `$item` pattern is different from `unix`.

Here's the solution that I applied for a project and you probably want to fix in the same way.

```PHP
use League\Flysystem\Adapter\Ftp as BaseFtp;
use League\Flysystem\AdapterInterface;

class Ftp extends BaseFtp
{
    protected function normalizeObject($item, $base)
    {
        if (stripos(ftp_systype($this->connection), 'windows') !== false) {
            return $this->normalizeWindowsObject($item, $base);
        }

        return parent::normalizeObject($item, $base);
    }

    protected function normalizeWindowsObject($item, $base)
    {
        $type = "file";

        if (strpos($item, '<DIR>') !== false) {
            $type = "dir";

            str_replace('<DIR>', '', $item);
        }

        $item = preg_replace('#\s+#', ' ', trim($item));

        list ($date, $time, $size, $name) = explode(' ', $item, 4);
        list($month, $day, $year) = explode('-', $date);

        $timestamp = strtotime($month . ' ' . $day . ' ' . $time);
        $path = empty($base) ? $name : $base . $this->separator . $name;

        if ($type === 'dir') {
            return compact('type', 'path', 'timestamp');
        }

        $visibility = AdapterInterface::VISIBILITY_PUBLIC;
        $size = (int) $size;

        return compact('type', 'path', 'visibility', 'size', 'timestamp');
    }
}
```

**Comment by frankdejonge** [comment]
This issue is a duplicate.

**Comment by frankdejonge** [comment]
This is a

**Comment by frankdejonge** [comment]
@kinncj could you supply some raw responses, I could create a windows ftp class (extended from the normal ftp) and perhaps eliminate this issue.

**Comment by kinncj** [comment]
` 2015-01-01 09:09:09  112233 file.extension`

 `2015-01-01 09:09:09 <DIR> 112233 directory_name`

where `112233` is the size

**Comment by frankdejonge** [comment]
@kinncj I see you're ignoring the year and default to public visibility. Any reasons for that? Is visibility not supported by windows ftp at all?

**Comment by kinncj** [comment]
not supported @frankdejonge 

it's

```
mm-dd-yy hh:mm  type  size file/dir
```

Be aware that: if it's a directory it sends <DIR>, otherwise it's empty.

That's why I check for <DIR> in the code above, cuz then I don't need to worry about empty spaces.

**Comment by frankdejonge** [comment]
@kinncj seems like there's pieces of your comment missing?

**Comment by frederikbosch** [comment]
We are also experiencing issues with Microsoft FTP Service. PHP errors on (https://github.com/thephpleague/flysystem/blob/master/src/Adapter/AbstractFtpAdapter.php#L289)[line 289 in AbstractFtpAdapter.php]. This is due to the absence of permissions.

```
Undefined offset: 4 in /vendor/league/flysystem/src/Adapter/AbstractFtpAdapter.php on line 289
Undefined offset: 8 in /vendor/league/flysystem/src/Adapter/AbstractFtpAdapter.php on line 289
```

Since, the errors are notice errors, the script continues and in our case continues without any other problems.

**Comment by barryvdh** [comment]
@frederikbosch Please post you raw responses also.

**Comment by frederikbosch** [comment]
@barryvdh Is this alright?

```
Response:	250 CWD command successful.
Command:	PWD
Response:	257 "/DG" is current directory.
Command:	PASV
Response:	227 Entering Passive Mode (000,000,000,00,000,00).
Command:	LIST
Response:	150 Opening BINARY mode data connection.
Response:	226 Transfer complete.
Listing:	03-25-15  10:21AM       <DIR>          IN
Listing:	03-25-15  10:44AM       <DIR>          UIT
Status:	Directory listing successful
Status:	Retrieving directory listing...
Command:	CWD /DG/IN
Response:	250 CWD command successful.
Command:	PASV
Response:	227 Entering Passive Mode (000,000,000,00,000,00).
Command:	LIST
Response:	150 Opening BINARY mode data connection.
Response:	226 Transfer complete.
Listing:	03-25-15  10:32AM                71046 ****.xml
Status:	Directory listing successful
```

**Comment by frankdejonge** [comment]
@frederikbosch is this the raw response of multiple calls? Or just that single rawlist call?

**Comment by frederikbosch** [comment]
@frankdejonge This is the response I logged from Filezilla. It contains two calls. One call with two directories (IN en UIT). The second call with one file (the xml file).

**Comment by barryvdh** [comment]
I came across this c library: http://cr.yp.to/ftpparse.html
The .c file also shows some different responses and how he handles it (in c code). They refer to this kind as MSDOS format. Maybe it helps a bit.

    /* UNIX-style listing, without inum and without blocks */
    /* "-rw-r--r--   1 root     other        531 Jan 29 03:26 README" */
    /* "dr-xr-xr-x   2 root     other        512 Apr  8  1994 etc" */
    /* "dr-xr-xr-x   2 root     512 Apr  8  1994 etc" */
    /* "lrwxrwxrwx   1 root     other          7 Jan 25 00:17 bin -> usr/bin" */
    /* Also produced by Microsoft's FTP servers for Windows: */
    /* "----------   1 owner    group         1803128 Jul 10 10:18 ls-lR.Z" */
    /* "d---------   1 owner    group               0 May  9 19:45 Softlib" */
    /* Also WFTPD for MSDOS: */
    /* "-rwxrwxrwx   1 noone    nogroup      322 Aug 19  1996 message.ftp" */
    /* Also NetWare: */
    /* "d [R----F--] supervisor            512       Jan 16 18:53    login" */
    /* "- [R----F--] rhesus             214059       Oct 20 15:27    cx.exe" */
    /* Also NetPresenz for the Mac: */
    /* "-------r--         326  1391972  1392298 Nov 22  1995 MegaPhone.sit" */
    /* "drwxrwxr-x               folder        2 May 10  1996 network" */
    
    /* MultiNet (some spaces removed from examples) */
    /* "00README.TXT;1      2 30-DEC-1996 17:44 [SYSTEM] (RWED,RWED,RE,RE)" */
    /* "CORE.DIR;1          1  8-SEP-1996 16:09 [SYSTEM] (RWE,RWE,RE,RE)" */
    /* and non-MutliNet VMS: */
    /* "CII-MANUAL.TEX;1  213/216  29-JAN-1996 03:33:12  [ANONYMOU,ANONYMOUS]   (RWED,RWED,,)" */
    
    /* MSDOS format */
    /* 04-27-00  09:09PM       <DIR>          licensed */
    /* 07-18-00  10:16AM       <DIR>          pub */
    /* 04-14-00  03:47PM                  589 readme.htm */
    
      /* Some useless lines, safely ignored: */
      /* "Total of 11 Files, 10966 Blocks." (VMS) */
      /* "total 14786" (UNIX) */
      /* "DISK$ANONFTP:[ANONYMOUS]" (VMS) */
      /* "Directory DISK$PCSA:[ANONYM]" (VMS) */

**Comment by frederikbosch** [comment]
@frankdejonge The adapter below fixes my Windows FTP problem. How can we convert it into a PR? Do you it to be build in into the Ftp/AbstractFtpAdapter? Or do you prefer a WindowsFtp adapter (outside the Flysystem package)?

```php
<?php
namespace MyVendor\Space;

use League\Flysystem\Adapter\Ftp;
use League\Flysystem\AdapterInterface;

class WindowsFtp extends Ftp {

    /**
     * Normalize a file entry.
     *
     * @param string $item
     * @param string $base
     *
     * @return array normalized file array
     */
    protected function normalizeObject($item, $base)
    {
        $item = preg_replace('#\s+#', ' ', trim($item), 7);
        list(/* $date */, /* $time */, $size, $name) = explode(' ', $item, 4);
        $permissions = '';

        $type = $this->detectType($permissions);
        $path = empty($base) ? $name : $base.$this->separator.$name;

        if ($type === 'dir') {
            return compact('type', 'path');
        }

        $permissions = $this->normalizePermissions($permissions);
        $visibility = $permissions & 0044 ? AdapterInterface::VISIBILITY_PUBLIC : AdapterInterface::VISIBILITY_PRIVATE;
        $size = (int) $size;

        return compact('type', 'path', 'visibility', 'size');
    }


}
```

**Comment by frederikbosch** [comment]
Never mind my previous reply, I forgot someone posted a (better) solution.

**Comment by jelrikvh** [comment]
:+1: For this issue. One of our data providers suddenly started using a Microsoft FTP server and I expected Flysystem would take care of the differences in return strings. I'm now using the implementation proposed in TS for the moment but would very much like to return to Vanilla Flysystem.

**Comment by JLeitchSCO** [comment]
As comment above. I primarily connect to Linux based FTP sites but recently had to connect to Windows FTP service and certain calls would not work as expected suc as listContents() and getSize().

Script provided by kinncj is so far working successfully (Thanks for that), but would prefer to have this built into the package going forward (fingers crossed)

**Comment by frankdejonge** [comment]
@jelrikvh are you working on that PR yet? ;)

**Comment by kinncj** [comment]
Can't see any open PR related to this, @jelrikvh.

**Comment by barryvdh** [comment]
Can you guys checkout https://github.com/thephpleague/flysystem/pull/487 ?

**Comment by frederikbosch** [comment]
@barryvdh Awesome, great work. Will let you know my results soon.

**Comment by frankdejonge** [comment]
This is resolved by #487
